Source/WebCore/page/text-extraction/TextExtraction.cpp

```diff
@@ -143,30 +143,54 @@ static inline std::optional<ItemData> extractItemData(Node& node, const HashMap<
             return { { ScrollableItemData { layer->scrollableArea()->totalContentsSize() } } };
     }
 
+    if (element->hasTagName(HTMLNames::olTag) || element->hasTagName(HTMLNames::ulTag))
+        return { { ContainerType::List } };
+
+    if (element->hasTagName(HTMLNames::liTag))
+        return { { ContainerType::ListItem } };
+
+    if (element->hasTagName(HTMLNames::blockquoteTag))
+        return { { ContainerType::BlockQuote } };
+
+    if (element->hasTagName(HTMLNames::articleTag))
+        return { { ContainerType::Article } };
+
+    if (element->hasTagName(HTMLNames::sectionTag))
+        return { { ContainerType::Section } };
+
+    if (element->hasTagName(HTMLNames::navTag))
+        return { { ContainerType::Nav } };
+
+    if (element->isLink())
+        return { { ContainerType::Link } };
+
     if (renderer->style().hasViewportConstrainedPosition())
         return { { ContainerType::ViewportConstrained } };
 
     return std::nullopt;
 }
 
-static inline void extractRecursive(Node& node, Item& parentItem, const HashMap<Ref<Text>, String>& extractedText)
+static inline void extractRecursive(Node& node, Item& parentItem, const HashMap<Ref<Text>, String>& extractedText, const std::optional<WebCore::FloatRect>& rectInRootView)
 {
     std::optional<Item> item;
-    if (auto itemData = extractItemData(node, extractedText))
-        item = { { WTFMove(*itemData), rootViewBounds(node), { } } };
+    if (auto itemData = extractItemData(node, extractedText)) {
+        auto nodeBoundsInRootView = rootViewBounds(node);
```

Comment:
I didn't look carefully at where this comes from, but do note that many kinds of "rectangle for this element" computations don't strictly contain their children

Reply:
(and thus are not safe to use to bail from recursion)

Reply:
Indeed! That said, in the current logic here, we'll continue recursing down anyways though since `itemData` will be `nullopt`, so if any of the children intersect with the target bounds, we'll be sure to include those. We _may_ want to preserve these "skipped" nodes in that case, though (tbh) I'm not sure.

I'll start with this for now, thanks for the reviews!